### DISCUSSION TITLE
Expanding decorator API

### DISCUSSION THREAD

**Discussion by manzt** [manzt/anywidget#295] [Ideas]
We added in `widget` decorator in #126.

This is still a little clunky to use as it adds several imports:

```python
from anywidget.experimental import widget
from psygnal import evented
from dataclasses import dataclass

@widget(esm="index.js")
@evented
@dataclass
class Counter:
    value: int = 0
```

It would be nice if `widget` could be smarter and reduced to something like:

```python
@widget(esm="index.js")
class Counter:
    value: int = 0
```

but the point of the descriptor API to to allow flexibility for modeling/serializing data and emitting events, so I'm not sure I want this to be too magical/opinionated.

I guess the use case I am really interested in is prototyping and avoiding imports. I'd like to have one import and get away with type-hints only if possible. Some prior art, but maybe we can just add a `dataclass` API of our own like pydantic.

```python
from anywidget.experimental import dataclass

@dataclass(esm="index.js")
class Counter:
    value: int = 0
```

Not sure if you have any thoughts @tlambert03

**Comment by manzt** [comment]
Alternatively, imports are fine and I shouldn't worry about this :)

**Comment by tlambert03** [comment]
yep, it's possible, and I do something similar in magicgui's `@guiclass` decorator actually:

https://github.com/pyapp-kit/magicgui/blob/5e068f31eaeeb130f43c38727b25423cc3ea4de3/src/magicgui/schema/_guiclass.py#L145-L162

you'll note there, however, all of the extra `@__dataclass_transform__` boilerplate that becomes required if you wish to preserve all of the nice static type hinting that one normally gets with a dataclass.  So, relatively easy to do, but one needs to be a bit careful with the implementation so as not to kill all the types

**Comment by manzt** [comment]
Thanks for the hint :) BTW, is it possible to propagate the type annotations for `evented` if the namespace is set to be something like events?

EDIT: just seeing https://psygnal.readthedocs.io/en/latest/dataclasses/#type-annotating-evented-dataclasses

**Comment by manzt** [comment]
Second, I'm wondering about how we could annotate fields for custom serialization/deserialization? Would this be something like `typing.Annotated`?

```python
from anywidget.experimental import dataclass
import ipywidgets

from typing import Annotated

@dataclass(esm="index.js")
class LayoutWidget:
    children: Annotated[list[ipywidgets.Widget], ipywidgets.widget_serialization]
```

Or somehow overload `Field`? Thinking I'll open a second issue for this...
